kubernetes_platform/python/kfp/kubernetes/toleration.py

```diff
@@ -0,0 +1,52 @@
+# Copyright 2023 The Kubeflow Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List
+
+from google.protobuf import json_format
+from kfp.dsl import PipelineTask
+from kfp.kubernetes import common
+from kfp.kubernetes import kubernetes_executor_config_pb2 as pb
+from kubernetes.client import V1Toleration
+
+
+def add_tolerations(
+    task: PipelineTask,
+    tolerations: List[V1Toleration],
+) -> PipelineTask:
+    """Add `tolerations<https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/>`_. to a task.
+
+    Args:
+        task: Pipeline task.
+        tolerations: A list of V1Tolerations defined using the Kubernetes Python Client.
+
+    Returns:
+        Task object with added tolerations.
+    """
+    tolerations_pb = []
+    for toleration in tolerations:
```

Comment:
I see in the unit tests that not all fields are required. Should we add validation for the optional/required fields?

One benefit of parameterizing with Python primitives is we can encode this information directly in the function interface.

Reply:
In the Python Kubernetes client, **none** of the fields are required!

```python
from kubernetes.client import V1Toleration

toleration = V1Toleration()
toleration
```

Output:

```python
{'effect': None,
 'key': None,
 'operator': None,
 'toleration_seconds': None,
 'value': None}
```

In fact, even on the backend, **none** of the fields are required (even the `key` field)! There are a few edge cases, handled [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L3900), where if one field is set to a specific value, other fields need to be set to specific values, but no single field is required in all cases.

In light of that, and considering the complexity of the validation, I suggest we just let the K8s API handle the validation, much like we did in V1.

Here's the code we used to test a bunch of different variations of toleration fields if you want to try it out yourself:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
  labels:
    env: test
spec:
  containers:
  - name: nginx
    image: nginx
    imagePullPolicy: IfNotPresent
  tolerations:
  - key: "example-key"
    operator: "Equal"
    effect: "NoSchedule"
    tolerationSeconds: 10
```

Reply:
What is the design rationale behind this approach? Is it calling for a pydantic usage?